agent/consul/state/catalog.go

```diff
@@ -3142,32 +3146,45 @@ func (s *Store) ServiceTopology(
 
 // combinedServiceNodesTxn returns typical and connect endpoints for a list of services.
 // This enabled aggregating checks statuses across both.
-func (s *Store) combinedServiceNodesTxn(tx ReadTxn, ws memdb.WatchSet, names []structs.ServiceName) (uint64, structs.CheckServiceNodes, error) {
+// Returns a slice of ServiceNames not associated with a ServiceNode instance.
+func (s *Store) combinedServiceNodesTxn(tx ReadTxn, ws memdb.WatchSet, names []structs.ServiceName) (uint64, structs.CheckServiceNodes, []structs.ServiceName, error) {
 	var (
-		maxIdx uint64
-		resp   structs.CheckServiceNodes
+		maxIdx   uint64
+		resp     structs.CheckServiceNodes
+		notFound []structs.ServiceName
```

Comment:
It would probably be preferable to move these checks for `notFound` services up to the ServiceTopology function, rather than returning them here:

- `combinedServiceNodesTxn` is called by our most heavily used endpoints (`/health`), and this new return value adds some overhead for a single less-common use-case.
- We should aim to not return more than 3 values from a function.

Reply:
Sounds good! I wanted to avoid a `N ServiceNames * n CheckServiceNodes` operation to find unused ServiceNames by doing it in one pass here but it does clutter the func signature a bit.

Reply:
re: `combinedServiceNodesTxn` being used by `/health`, I only see this func being used in ServiceTopology-related endpoints. Is there an RPC call I'm missing?

Reply:
Ah you're right. I got this function confused with `parseCheckServiceNodes`. 

I think what you have now looks good though.